Add a route provider and controller for add page/form.

bojanz opened the pull request:
Continues the work from #7:

The controller has been rewritten. Addressed feedback:
- Deleted the collections route generation.
- Added a custom theme for the add list.
- Added the missing redirect&access checks to the addPage method.

bojanz committed:
Add a route provider and controller for add page/form.

dawehner commented on src/Controller/EntityCreateController.php:
```diff
@@ -0,0 +1,126 @@
+    $entity_type = $this->entityTypeManager()->getDefinition($entity_type_id);
+    $bundle_type = $entity_type->getBundleEntityType();
+    $form_route_name = 'entity.' . $bundle_type . '.add_form';
+    $bundles = $this->entityTypeManager()->getStorage($bundle_type)->loadMultiple();
```
We need to use \Drupal\Core\Entity\EntityTypeBundleInfoInterface instead of loading the bundles directly.

dawehner commented on src/Controller/EntityCreateController.php:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    return [
+      '#theme' => 'entity_add_list',
```
We should probably add some cache list tag for the bundle entity type?

dawehner commented on src/Controller/EntityCreateController.php:
```diff
@@ -0,0 +1,126 @@
+   */
+  public function addPageTitle($entity_type_id) {
+    $entity_type = $this->entityTypeManager()->getDefinition($entity_type_id);
+    $bundle_type = $entity_type->getBundleEntityType();
```
same here

dawehner commented on src/Controller/EntityCreateController.php:
```diff
@@ -0,0 +1,126 @@
+
+    $args = [
+      '@entity-label' => $entity_type->getLabel(),
+      '%type' => $bundle ? $bundle->label() : '',
```
what about using %bundle instead as placeholder?

dawehner commented on src/Routing/CreateUIRouteProvider.php:
```diff
@@ -0,0 +1,79 @@
+  protected function addPageRoute(EntityTypeInterface $entity_type) {
+    if ($entity_type->hasLinkTemplate('add-page')) {
+      $route = new Route($entity_type->getLinkTemplate('add-page'));
+      $route->setDefault('_controller', '\Drupal\src\Controller\EntityCreateController::addPage');
```
Note: Totally broken classnames

dawehner commented on src/Routing/CreateUIRouteProvider.php:
```diff
@@ -0,0 +1,79 @@
+  protected function addFormRoute(EntityTypeInterface $entity_type) {
+    if ($entity_type->hasLinkTemplate('add-form')) {
+      $route = new Route('entity.' . $entity_type->id() . '.add-form');
+      $route->setDefault('_controller', '\Drupal\src\Controller\EntityCreateController::addForm');
```
here also

bojanz commented:
Note that I moved back to loading bundle entities (from using the bundle info service) cause we need them anyway for the listing. That's what NodeController does as well.

bojanz committed:
Rename the create route provider, add a matching admin route provider.

bojanz committed:
Make the add form usable on entities without bundles (such as node types).

dawehner commented on entity.module:
```diff
@@ -0,0 +1,55 @@
+ *   - form_route_name: The add form route.
+ */
+function template_preprocess_entity_add_list(&$variables) {
+  $bundle_type = \Drupal::service('entity_type.manager')->getDefinition($variables['bundle_type']);
```
There is \Drupal::entityTypeManager()

dawehner commented on entity.module:
```diff
@@ -0,0 +1,55 @@
+function template_preprocess_entity_add_list(&$variables) {
+  $bundle_type = \Drupal::service('entity_type.manager')->getDefinition($variables['bundle_type']);
+  $variables += [
+    'class' => str_replace('_', '-', $bundle_type->getBundleOf()) . '-add-list',
```
that is sooo d7, classes should be nowhere, and just be added by some custom template. At least this is what code is doing.

dawehner commented on entity.module:
```diff
@@ -0,0 +1,55 @@
+  $variables += [
+    'class' => str_replace('_', '-', $bundle_type->getBundleOf()) . '-add-list',
+    'create_bundle_url' => Url::fromRoute('entity.' . $bundle_type->id() . '.add_form'),
+    'bundle_type_label' => $bundle_type->getLowercaseLabel(),
```
That feels kinda wrong

dawehner commented on entity.module:
```diff
@@ -0,0 +1,55 @@
+    $url = Url::fromRoute($variables['form_route_name'], [$bundle->getEntityTypeId() => $bundle->id()]);
+    $variables['bundles'][$bundle->id()] = [
+      'bundle' => $bundle->id(),
+      'add_link' => \Drupal::l($bundle->label(), $url),
```
Let's use a link object instead.

dawehner commented on src/Controller/EntityCreateController.php:
```diff
@@ -0,0 +1,170 @@
+      if (!$access->isAllowed()) {
+        unset($bundles[$index]);
+      }
+      $this->renderer->addCacheableDependency($build, $access);
```
You don't need that, just add #access to the render array.

bojanz replied:
Do you think we should move this entire segment from preprocess to the controller:
```php
  foreach ($variables['bundles'] as $bundle) {
    $variables['bundles'][$bundle->id()] = [
      'bundle' => $bundle->id(),
      'add_link' => Link::createFromRoute($bundle->label(), $variables['form_route_name'], [$bundle_type => $bundle->id()]),
      'description' => [
        '#markup' => $bundle->getDescription(),
      ],
    ];
  }
```
Felt cleaner to do it the way it is right now.

bojanz replied:
We'd need a render array of bundle info (instead of expanding it in the preprocess function), but what's the type of each element?

dawehner replied:
I'm a little bit confused about that and don't understand the question

dawehner commented on src/Controller/EntityCreateController.php:
```diff
@@ -0,0 +1,170 @@
+        'tags' => $entity_type->getListCacheTags(),
+      ],
+      '#bundle_type' => $bundle_type,
+      '#form_route_name' => $form_route_name,
```
I'd have expected to create the URL objects here, so we don't have to do the thing in the template layer

dawehner commented on src/Controller/EntityCreateController.php:
```diff
@@ -0,0 +1,170 @@
+      $bundle_name = reset($bundles);
+      return $this->redirect($form_route_name, [$bundle_type => $bundle_name]);
+    }
+    // The theme function needs the full bundle entities.
```
... bundles are still definable as something else than config entities. IMHO we should support that, and make some of the things optional around bundles, where we have more information

dawehner commented on src/Routing/AdminCreateHtmlRouteProvider.php:
```diff
@@ -0,0 +1,39 @@
+ *
+ * @see \Drupal\entity\Routing\CreateHtmlRouteProvider.
+ */
+class AdminCreateHtmlRouteProvider extends CreateHtmlRouteProvider {
```
good idea, IMHO the @see is a bit pointless because its just pointing to the parent class

dawehner commented on templates/entity-add-list.html.twig:
```diff
@@ -0,0 +1,32 @@
+ */
+#}
+{% if bundles is not empty %}
+  <dl class="entity-add-list {{ class }}">
```
yeah we should have an attributes object + define the classes at the top of the file

bojanz committed:
Addressed feedback.

- removed the classes to match core behavior.
- started using the link object.
- Added an interface for bundle entity descriptions.
- dawehner's fix for create_bundle_url in the preprocess function.

bojanz commented:
Addressed the smaller feedback (rebase + latest commit). 
Introduced an interface for getDescription(), can't use $bundle->getEntityKey('description') cause it's protected.

What remains is to move the bundle list creation to the controller (from the preprocess function).
I'd leave the expansion to support non-config-entity bundles to another issue, cause it's an edge case that will require additional effort.

dawehner commented on src/Entity/BundleConfigEntityInterface.php:
```diff
@@ -0,0 +1,35 @@
+/**
+ * Defines the interface for config entities that act as bundles.
+ */
+interface BundleConfigEntityInterface extends ConfigEntityInterface {
```
I would have expected an entity key to be honest

bojanz commented:
Renamed the interface from BundleConfigEntityInterface to EntityDescriptionInterface.

Done for now, waiting for Daniel to demonstrate how he'd like the renderable array / preprocess cleanup to look.

bojanz committed:
Support bundles that aren't provided by entity types, add tests.

dawehner commented:
:ship: beside the tests

bojanz committed:
Expand the add page test.
